auparse/auparse_test.go

```diff
@@ -123,41 +142,176 @@ func TestExtractKeyValuePairs(t *testing.T) {
+			},
+		},
+		{
+			"short message",
+			`x='grep "test" file' y=z`,
+			fieldMap{
+				"x": {`'grep "test" file'`, `grep "test" file`},
+				"y": newField("z"),
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		b.Run(tc.name, func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				extractKeyValuePairs(tc.in)
+			}
+		})
+	}
+}
+
+var dataTests = []struct {
+	name   string
+	text   string
+	output map[string]string
+}{
+	{
+		name: "syscall",
+		text: "type=SYSCALL msg=audit(1481076984.827:17): arch=c000003e syscall=313 success=yes exit=0 a0=0 a1=41a15c a2=0 a3=0 items=0 ppid=390 pid=391 auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm=\"modprobe\" exe=\"/usr/bin/kmod\" subj=system_u:system_r:insmod_t:s0 key=(null)",
+		output: map[string]string{
+			"a0":          "0",
+			"a1":          "41a15c",
+			"a2":          "0",
+			"a3":          "0",
+			"arch":        "x86_64",
+			"auid":        "unset",
+			"comm":        "modprobe",
+			"egid":        "0",
+			"euid":        "0",
+			"exe":         "/usr/bin/kmod",
+			"exit":        "0",
+			"fsgid":       "0",
+			"fsuid":       "0",
+			"gid":         "0",
+			"items":       "0",
+			"pid":         "391",
+			"ppid":        "390",
+			"result":      "success",
+			"ses":         "unset",
+			"sgid":        "0",
+			"subj_domain": "insmod_t",
+			"subj_level":  "s0",
+			"subj_role":   "system_r",
+			"subj_user":   "system_u",
+			"suid":        "0",
+			"syscall":     "finit_module",
+			"tty":         "(none)",
+			"uid":         "0",
+		},
+	},
+	{
+		"user_cmd",
+		`type=USER_CMD msg=audit(1488862769.030:19469538): user pid=3027 uid=497 auid=700 ses=11988 msg='cwd="/" cmd=2F7573722F6C696236342F6E6167696F732F706C7567696E732F636865636B5F617374657269736B5F7369705F7065657273202D7020313037 terminal=? res=success'`,
+		map[string]string{
+			"auid":   "700",
+			"cmd":    "/usr/lib64/nagios/plugins/check_asterisk_sip_peers -p 107",
+			"cwd":    "/",
+			"pid":    "3027",
+			"result": "success",
+			"ses":    "11988",
+			"uid":    "497",
+		},
+	},
+}
+
+func TestExtractAndEnrichData(t *testing.T) {
+	tests := dataTests
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			msg, err := ParseLogLine(tc.text)
+			if err != nil {
+				t.Fatalf("parsing message: %v", err)
+			}
+			data, err := msg.Data()
+			require.Nil(t, err)
+			assert.Equal(t, tc.output, data)
+		})
+	}
+}
+
+func BenchmarkAuditMessageData(b *testing.B) {
+	tests := dataTests
+	for _, tc := range tests {
+		tc := tc
+		b.Run(tc.name, func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				b.StopTimer()
+				var (
+					data map[string]string
+					err  error
+				)
+				msg, errP := ParseLogLine(tc.text)
+				if errP != nil {
+					b.Fatalf("parsing message: %v", errP)
+				}
+				b.StartTimer()
+				data, err = msg.Data()
+				b.StopTimer()
+				require.Nil(b, err)
+				assert.Equal(b, tc.output, data)
+			}
+		})
```

Comment:
The `ParseLogLine` call can be hoisted out of the loop and the check probably should be in a test rather than in the benchmark, just leaving a nil check in the benchmark to avoid the compiler optimising it away.

```suggestion
		msg, err := ParseLogLine(tc.text)
		if err != nil {
			b.Errorf("parsing message: %v", err)
			continue
		}
		b.Run(tc.name, func(b *testing.B) {
			for i := 0; i < b.N; i++ {
				_, err = msg.Data()
				if err != nil {
					b.Fatalf("unexpected error: %v", err)
				}
			}
		})
```

Reply:
Thanks for the suggestion 👍🏼 looks nicer

Reply:
Actually I noticed this output after applying your suggestion, it does not seem correct

```
BenchmarkAuditMessageData/syscall-8                 	1000000000	         2.342 ns/op	       0 B/op	       0 allocs/op
BenchmarkAuditMessageData/user_cmd-8                	1000000000	         2.352 ns/op	       0 B/op	       0 allocs/op
```

Not sure what's happening: could it be that we don't make any use of the map returned by `Data()`, and that makes it not part of the calculation of the benchmark time/memory?

Reply:
I had to revert your suggestion, the previous form is the only one that allows the benchmark tests to run and return meaningful results.

Reply:
> Not sure what's happening: could it be that we don't make any use of the map returned by Data(), and that makes it not part of the calculation of the benchmark time/memory?

The SSA passes won't omit part of the return, examining the `err` should be enough to ensure that the code is run. What happens if the err is declared in the loop? What happens if you only hoist out of the loop and not out of the closure?